### PR TITLE
compute: add diff result type

### PR DESCRIPTION
**Pull request by rvantonder**
Add dedicated diff result type. It is currently only used/generated by compute. I implemented the result methods even though I don't need them, and they're not used yet, so that in time we should migrate to using this diff result type and not proxy via `CommitMatch`.

Thanks @camdencheek for the diff parsing code in https://github.com/sourcegraph/sourcegraph/tree/cc/diff-parser!

## Test plan
Added unit tests.

### CODE REVIEW

File: internal/search/result/commit_diff.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+	"github.com/sourcegraph/sourcegraph/internal/gitserver/gitdomain"
+	"github.com/sourcegraph/sourcegraph/internal/search/filter"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+)
+
+type CommitDiffMatch struct {
+	Commit  gitdomain.Commit
+	Repo    types.MinimalRepo
+	Preview *MatchedString
```
Comment (rvantonder) [inline internal/search/result/commit_diff.go:17]: Even though this value isn't used by compute use case, it will let us migrate away more easily from `CommitMatch`.

File: internal/search/result/commit_diff_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     font-weight: normal;
++    display: inline !important;
+ }
+`
+
+	res, err := ParseDiffString(input)
+	if err != nil {
+		panic(err)
+	}
+	autogold.Equal(t, res)
```
Comment (camdencheek) [inline internal/search/result/commit_diff_test.go:48]: Could we flesh out these tests a bit before merging? I wrote this parser in like 10 minutes 😬

File: internal/search/result/match.go [hunk 3]
```diff
@@ -73,6 +74,11 @@ type Key struct {
 	// Empty if there is no file associated with the match (e.g. RepoMatch or CommitMatch)
 	Path string
 
+	// PathMode is whether a file path has been created, modified or
+	// removed. Applies to, e.g., CommitDiffMatch. Empty if no there is no
+	// associated path mode
+	PathStatus PathStatus
```
Comment (camdencheek) [inline internal/search/result/match.go:80]: Why is this added to `Key`? Is it needed for sorting or dedup?

File: internal/search/result/commit_diff_test.go (hunk 2 above)
Comment (camdencheek) [reply internal/search/result/commit_diff_test.go:48]: Maybe a good candidate for the new builtin fuzz testing 👀

Comment (camdencheek) [reply internal/search/result/commit_diff_test.go:48]: I fuzzed with this and found no panics, so 🤷 

```go
func FuzzParseDiffString(f *testing.F) {
	f.Add("file1 file2\n@@ -1,2 +1,6 @@ ... +1\n a\n+b\n-c\n d")
	f.Fuzz(func(t *testing.T, input string) {
		ParseDiffString(input)
	})
}
```

File: internal/search/result/commit_diff.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+	*DiffFile
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply internal/search/result/commit_diff.go:18]: For this type, I'd ideally want/expect `DiffFile` representation to be the definitive one (`Preview` should not exist, only `Ranges`).

I can delete `Preview` if ya want, then there's no multiple different representation. Going ahead with merge tho.

File: internal/search/result/match.go (hunk 3 above)
Comment (rvantonder) [reply internal/search/result/match.go:80]: thought I needed it, think I don't, removed it

File: internal/search/result/commit_diff.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+	newCount, err = strconv.Atoi(groups[4])
+	if err != nil {
+		return 0, 0, 0, 0, "", err
+	}
+	return oldStart, oldCount, newStart, newCount, groups[5], nil
+}
+
+type DiffFile struct {
+	OrigName, NewName string
+	PathStatus        PathStatus
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply internal/search/result/commit_diff.go:202]: yeah good call
